### ISSUE TITLE
training data

### ISSUE THREAD

**Issue by 41924076** [HazyResearch/m2#27]
hi, thank you for your nice work! Do you train your M2-BERT-128 ~ 32K (shown in the paper) on LOCO V0 or LOCO V1 training set?

**Comment by DanFu09** [comment]
The public checkpoints are trained on LoCoV0. New checkpoints on LoCoV1
will be coming out soon!

**Comment by 41924076** [comment]
thank you for your reply!

**Comment by jonsaadfalcon** [comment]
We use random negatives for fine-tuning. At each step for OPL, we calculate the similarity between the query and the positive passage as well as the similarity between the query and the negative passage.

**Comment by 41924076** [comment]
thank you for your reply!

**Comment by 41924076** [comment]
Hello, I find that it's a little bit hard to make qmsum score more than 45, do you use the whole LOCO V0 as training set with no duplication, deletion or certain data proportion? : )

**Comment by jonsaadfalcon** [comment]
We use all of LoCo V0 as our fine-tuning dataset with 32 negatives for every query-positive passage pair. The public checkpoints only go through pretraining and LoCoV0 fine-tuning. However, we plan to release an updated version of the QMSUM dataset (as well as several new datasets) in LoCoV1 soon!

**Comment by 41924076** [comment]
Thank you for your reply! Have a nice day!

**Comment by 41924076** [comment]
Hello, sorry to bother you. I'd like to try reproducing the training of m2bert on loco v0 or v1 with 2048 length. 

According to your information and the paper, m2bert loco v0 was trained with OPL loss. However, the main branch on GitHub does not provide a statement script for training (https://github.com/HazyResearch/m2/blob/main/bert/EMBEDDINGS.md#training), and the code does not include OPL loss.

Additionally, I noticed that the jonsf branch (https://github.com/HazyResearch/m2/tree/jonsf-patch-1) includes OPL loss, but I'm unsure if the gather_loco_training_example in the jonsf branch is suitable for OPL loss. The training script provided in the jonsf branch (https://github.com/HazyResearch/m2/blob/jonsf-patch-1/bert/EMBEDDINGS.md#training) also uses GradCache's multiple_negatives_ranking_loss instead of OPL. Can this script produce good results similar to those mentioned in the paper?

If it's convenient for you, could you please provide a OPL (or  multiple_negatives_ranking) loss training script that can reproduce public checkpoint?

**Comment by 41924076** [comment]
thank you for your reply!!!

**Comment by 41924076** [comment]
hi, i tried to train m2bert on loco v0. i use the m2bert-2k-pretrained-checkpoint and training_yaml=yamls/finetune-glue/embedding_training.yaml. 
I trained with tevatron's loss, 4neg per query, batchsize=1, 8gpu, and use negatives_x_devices, which means one query have 32 random neg in total. I trained 4 epochs and get the result as follows:

Summ ScreenFD 0.74
Gov.Report 0.92
QMSUM 0.46
QASPER Title 0.64
QASPER Abstract 0.81

avg 71.4

Do you think it's a correct result?

**Comment by sorenmc** [comment]
Hi, thank you for publishing your results and sharing your training code! 
It looks like you are importing OPL as `sentence_transformers.losses.OrthogonalPassageLoss`.
However when i check sentence transformers they don't have this loss, so I'm assuming this is a fork of sentence_transformers, that you haven't shared?

From above and your paper it sounds like you are doing something like below pseduo_code

```python
def opl_loss(model, query, documents, labels):
    q_embedding = model(query)
    d_embeddings = model(documents) # 1 positive + 32 randomly sampled negatives
    pairwise_cosine_sim = cosine_sim(q_embedding, d_embeddings)
    loss = mse(pairwise_cosine_sim, labels)
    return loss
```

Does this sound about right?

**Comment by jonsaadfalcon** [comment]
Yes, we have a custom version of the SentenceTransformers code base, in which we add orthogonal projection loss (OPL). We include the instructions for importing it in [the M2 codebase](https://github.com/HazyResearch/m2/blob/main/bert/EMBEDDINGS.md) but here is [the link to the codebase](https://github.com/jonsaadfalcon/sentence-transformers).

**Comment by sorenmc** [comment]
Thank you for the quicky reply. This is exactly what I was looking for.
